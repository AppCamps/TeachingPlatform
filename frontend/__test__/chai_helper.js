import chai, { expect } from "chai";
import chaiEnzyme from "chai-enzyme";

chai.use(chaiEnzyme());
chai.config.truncateThreshold = 0;
chai.config.showDiff = true;

export default chai;

export { expect };
