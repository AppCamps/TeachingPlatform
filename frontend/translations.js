export const translations = {
  'de': {
    'If the email is registered with App Camps and not confirmed yet you will soon receive a link to confirm it.': 'Falls die angegebene E-Mail-Adresse bei uns existiert bekommst du gleich einen Link zum Bestätigen deiner E-Mail-Adresse zugeschickt.',
    'Email confirmed. You may now login using your email and password.': 'Deine E-Mail-Adresse ist nun bestätigt. Du kannst dich jetzt mit deinem neuen Passwort anmelden.',
    'The confirmation of your email address was not successful. Please, contact us via philipp@appcamps.de so that we can confirm your account.': 'Die Bestätigung deiner E-Mail-Adresse war nicht erfolgreich. Bitte melde dich bei philipp@appcamps.de. Dann können wir deine E-Mail Adresse freischalten.',
    'Your confirmation link has expired. Please restart the confirmation process.': 'Dein Bestätigungslink ist abgelaufen. Bitte starte denn Prozess erneut.',
    'If the email you provided is registered with App Camps you will soon receive an email containing a link to the reset form.': 'Falls die angegebene E-Mail-Adresse bei uns existiert bekommst du gleich einen Link zum zurücksetzen deines Passworts zugeschickt.',
    'Password successfully changed. You can now login with your new password.': 'Dein Passwort wurde erfolgreich geändert. Du kannst dich jetzt mit deinem neuen Passwort anmelden.',
    'There was an error with your reset link. Please restart the password recovery process.': 'Die Änderung deines Passworts war leider nicht erfolgreich. Bitte starte denn Prozess erneut.',
    'Your password reset link has expired. Please restart the password recovery process.': 'Die Zeit zum Ändern deines Passworts ist abgelaufen. Bitte starte denn Prozess erneut.',
    'Appcamps on {name}': 'Appcamps bei {name}',
    'With {love} from {brand}': 'Mit {love} von {brand}',
    'Imprint': 'Impressum',
    'Privacy Policy': 'Datenschutzerklärung',
    'facebook': 'Facebook',
    'twitter': 'Twitter',
    'Change address': 'Adresse ändern',
    'Edit school information': 'Schuldaten ändern',
    'Help': 'Hilfe',
    'Platform tips': 'Tipps zur Plattform',
    'Frequently asked questions': 'Häufige Fragen',
    'Contact team': 'Kontakt zum Team',
    'Certificates': 'Zertifikate',
    'Edit user information': 'Nutzerdaten ändern',
    'Edit password': 'Passwort ändern',
    'Logout': 'Abmelden',
    'This lesson was tagged as completed.': 'Diese Lektion wurde als erledigt makiert.',
    'Undo': 'Rückgangig machen',
    'Tag lesson as completed': 'Lektion als erledigt markieren',
    'back to my classes': 'zur Klassenübersicht',
    'You did not add any courses yet': 'Du hast noch keine Kurse hinzugefügt',
    'Edit class': 'Klasse bearbeiten',
    'Not yet started': 'Noch nicht gestartet',
    'As soon as you mark the first lession as completed, you will see in the progress bar how far are you in the class.': 'Sobald du die erste Lektion als erledigt markiert hast, siehst du am Balken, wie weit du mit der Klasse bist',
    'Continue course': 'Kurs fortsetzen',
    'Start course': 'Kurs starten',
    'Successfully concluded': 'Erfolgreich beendet!',
    'Show all classes': 'Alle Klassen anzeigen',
    'My classes': 'Meine Klassen',
    'Add a new class': 'Neue Klasse anlegen',
    'Why to add classes?': 'Warum Klassen anlegen?',
    'Helpful for you:': 'Hilfreich für dich:',
    'you can manage more classes and see their progress.': 'Du kannst mehrere Klassen verwalten und siehst den Fortschritt.',
    'Helpful for us:': 'Hilfreich für uns:',
    'through the existing classes we see how many students we could reach. When we show that we reach many students, we can keep our offer by financing ourselves via funding and providing support also in the future for free.': 'Durch angelegte Klassen erfahren wir, wie viele Schüler/innen wir erreichen. Es geht uns um die Anzahl der Schüler. Namen oder persönliche Daten fragen wir nicht ab. Wenn wir zeigen, dass wir viele Schüler erreichen, können wir unser Angebot über Stiftungen finanzieren und die Unterlagen auch in Zukunft kostenlos anbieten.',
    'back to course overview': 'zurück zum Kursangebot',
    'Course preview': 'Kursvorschau',
    'You did not save your data! Do you still want to leave the page?': 'Du hast deine Angaben nicht gespeichert. Willst du die Seite trotzdem verlassen?',
    '(You can save the class in the third step, please scroll down.)': '(Die Klasse kannst du im dritten Schritt anlegen, bitte runter scrollen.)',
    'Create class': 'Klasse anlegen',
    'Create a new class': 'Neue Klasse anlegen',
    'cancel': 'Abbrechen',
    'Teaching material': 'Unterrichtsmaterial',
    'News': 'Neuigkeiten',
    'You did not save your changes! Do you still want to leave the page?': 'Du hast deine Änderungen nicht gespeichert. Willst du die Seite trotzdem verlassen?',
    '(You can save in the third step, please scroll down.)': '(Speichern kannst du im dritten Schritt, bitte runter scrollen.)',
    'Save': 'Speichern',
    '{email} is still unconfirmed. {emailConfirmationRequestLink}': '{email} ist noch nicht bestätigt. {emailConfirmationRequestLink}',
    'Resend confirmation link': 'Bestätigungslink erneut zusenden',
    'Saved': 'Gespeichert',
    'Email': 'E-Mail',
    'First name': 'Vorname',
    'Last name': 'Nachname',
    'Edit your data': 'Ändere deine Daten',
    'Current password': 'Aktuelles Passwort',
    'New password': 'Neues Passwort',
    'Confirm new password': 'Neues Passwort bestätigen',
    'Email confirmation': 'E-Mail bestätigen',
    'We will send you an email containing a link to confirm your email. If you have not received an email, yet, please contact us at {supportEmail}': 'Wir schicken dir eine E-Mail mit dem Bestätigungs-Link zu. Falls du keine E-Mail erhälst, wende dich bitte an {supportEmail}.',
    'Send': 'Absenden',
    'Validating email confirmation link...': 'Bestätigungslink wird überprüft...',
    'Information about certificates': 'Infos zu den Zertifikaten',
    'How can I download certificates?': 'Wie kann ich Zertifikate herunterladen?',
    'In this video you will learn how and where you can dowload the certificates for your pupils': 'Im Video erklären wir dir, wie und wo du Zertifikate für deine Schülerinnen und Schüler findest',
    'Tips for the usage of the platform': 'Tipps zur Nutzung der Plattform',
    'Are you new here or would you like to know how can you optimally work with the platform and the documents?': 'Du bist neu hier oder willst gerne wissen, wie du optimal mit unserer Plattform und den Unterlagen arbeiten kannst?',
    'In the video, we introduce the platform and give some usage tips': 'In dem Video stellen wir die Plattform vor und geben Tipps zur Nutzung',
    'Do you prefer to read? Here you can find the PDF': 'Du liest dir das lieber durch? Hier findest du ein PDF',
    'please, click': 'bitte anklicken',
    'Welcome to App Camps': 'Willkommen bei App Camps',
    'It is crucial for us to know, where our materials are used. Before you can setup your classes, we need you to tell us a bit more about you.': 'Für uns ist es wichtig zu wissen, wo unsere Unterlagen eingesetzt werden. Bitte fülle daher dieses kurze Formular aus.',
    'Invalid email or password': 'E-Mail oder Passwort ungültig',
    'Your session has expired': 'Deine Sitzung ist abgelaufen',
    'Unconfirmed email.': 'E-Mail-Adresse unbestätigt.',
    'Did not sign up yet? {registrationLink}': 'Du bist neu hier? {registrationLink}',
    'Register here': 'Hier geht’s zur Registrierung',
    'Did not receive a confirmation email? {confirmationLink}': 'Keine Bestätigungsmail erhalten? {confirmationLink}',
    'Click here': 'Hier klicken',
    'Login': 'Anmelden',
    'Forgot your password?': 'Passwort vergessen?',
    'Password': 'Passwort',
    'Code': 'Code',
    'To the card': 'Zu den Karten',
    'Here you go to the learning cards': 'So geht\'s zu den Lernkarten',
    'Open the URL in the browser': 'URL in den Browser eingeben',
    'Submit this code': 'Mit diesem Code freischalten',
    'Password reset': 'Passwort zurücksetzen',
    'We will send you an email containing a link to reset your password. If you have not received an email, yet, please contact us at {supportEmail}': 'Wir schicken dir eine E-Mail mit einem Link zum Zurücksetzen deines Passworts zu. Falls du keine E-Mail erhälst, wende dich bitte an {supportEmail}.',
    'Password confirmation': 'Passwort bestätigen',
    'previous posts': 'vorherige Beiträge',
    'next posts': 'weitere Beiträge',
    'Released on {releaseDate}': 'Veröffentlicht am {releaseDate}',
    'Soon we will provide interesting news from us and our partners here.': 'Bald findest du hier Neuigkeiten von uns und unseren Partnern.',
    'How did you hear from us?': 'Wie hast du von uns erfahren?',
    'Comment': 'Anmerkungen',
    'By registrating I accept the following terms:': 'Mit deiner Registrierung stimmst du den folgenden Bedingungen zu:',
    'I will not forward any data or teaching materials to third parties. My colleagues may register themselves on interest.': 'Ich gebe meine Anmeldedaten nicht an Dritte weiter. Meine Kollegen können sich bei Interesse ebenfalls registrieren.',
    'I accept to receive our newsletter per email (e.g notifications about new teaching materials). You may revoke this any time by clicking the provided links in the email or by sending an email to {email}.': ' Ich stimme zu über Neuigkeiten per E-Mail informiert zu werden (z.B. Ankündigung neue Unterlagen). Dazu nutzen wir die Dienste Mailchimp und Intercom. Die Zustimmung kann jederzeit per E-Mail an {email} widerrufen werden.',
    'I will answer questionnaires to evaluate the quality of our teaching materials. Before and after the execution of every course there will be short questionnaires (approx. 5 minutes). To be able to provide our platform for free we need to collect metrics to prove that our teaching materials are working.': 'Ich beteilige mich an den Fragebögen zur Evaluation der Unterlagen. Vor und nach jeder Durchführung eines Kurses gibt es kurze Fragebögen (ca. 5 Minuten). Damit wir unser Angebot weiterhin kostenlos anbieten können brauchen wir Zahlen die belegen, wie die Unterlagen eingesetzt werden.',
    'Data protection': 'Datenschutz',
    'I have read and accept the {privacyPolicy} of App Camps.': 'Ich habe die {privacyPolicy} von App Camps gelesen und stimme zu.',
    'Sign up free of charge': 'Jetzt kostenlos registrieren',
    'student': 'Schüler',
    'We are sorry, but this platform was built with teachers and course instructors in mind. Nonetheless we provide a newsletter for students like you, where we will keep you informed about programming, computer science and technology.': 'Unsere Plattform ist nur für Lehrkräfte und Kursleiter*innen. Wenn du als Schüler*in gerne eine AG oder einen Workshop für deine Mitschüler anbieten willst, dann schick uns eine E-Mail an {email}.',
    'Please read our blog post': 'Hier findest du einen Bericht über Fynn, der an seiner Schule eine AG mit unseren Unterrichtsmaterialien anbietet: {bloglink}',
    'To blog post': 'Zum Blogbeitrag',
    'parent': 'Eltern',
    'We are sorry, but this platform was built with teachers and course instructors in mind. Nonetheless we provide a newsletter for parents, where we will keep you informed about programming, computer science and technology.': 'Unsere Plattform ist nur für Lehrkräfte und Kursleiter*innen. Eltern sind herzlich eingeladen, Workshops mit unseren Unterrichtsmaterialien an Schulen oder außerschulisch anzubieten. Bitte registriere dich in diesem Fall als Kursleiter*in.',
    'Teacher': 'Lehrer*in',
    'Course instructor': 'Kursleiter*in',
    'Parent': 'ein Elternteil',
    'Student': 'Schüler*in',
    'Please sign up for free access': 'Registriere dich, um kostenlos Zugang zu erhalten',
    'Already registered? {signIn}': 'Bereits registriert? {signIn}',
    'Sign in': 'Hier gehts zur Anmeldung',
    'I am a': 'Ich bin',
    'Thank you for you interest in App Camps': 'Danke für dein Interesse an App Camps',
    'Subscribe to our {resourceName} newsletter': 'Abonniere unseren {resourceName}newsletter',
    'Subscribe': 'Abonnieren',
    'Registration success': 'Registrierung erfolgreich',
    'Welcome to App Camps, {firstName}!': 'Herzlich willkommen bei App Camps, {firstName}!',
    'You will soon receive an confirmation email. To complete your registration, please click on the confirmation link in the email. Afterwards you will be able to sign in.': 'Du wirst bald eine Bestätigungs-E-Mail erhalten. Klicke bitte auf den Link in der E-Mail um deine Registrierung abzuschließen. Danach kannst du dich einloggen.',
    'If the email does not arrive within the next minutes please {resendConfirmationLink} it or {supportEmailLink}.': 'Falls die E-Mail nicht innerhalb der nächsten Minuten ankommt {resendConfirmationLink} oder {supportEmailLink}.',
    'send us an email': 'melde dich bei uns',
    'click here to resend': 'lass sie dir erneut zuschicken',
    'In a lesson (duty)': 'Im regulären Unterricht (Pflichtbereich)',
    'In a lesson (choice)': 'Im regulären Unterricht (Wahlbereich)',
    'With a project group': 'In einer AG',
    'On project days': 'Während der Projekttage',
    'In which project group will you use the materials?': 'In welcher AG findet der Kurs statt?',
    'In what kind of project will you use the materials?': 'Um was für ein Projekt handelt es sich?',
    'In which school subject will you use the materials?': 'In welchem Fach findet der Kurs statt?',
    'Add information about your class': 'Angaben zu diesem Klassenprofil',
    'Class name': 'Name der Klasse',
    'School year': 'Schuljahr',
    'Grade': 'Klassenstufe',
    'Girl count': 'Anzahl der Mädchen',
    'Boy count': 'Anzahl der Jungen',
    'Where will you use the materials?': 'Einsatzrahmen',
    'Next': 'Weiter',
    'Back': 'Zurück',
    'No Courses found for topic {topic}': 'Für {topic} konnten keine Kurse gefunden werden',
    'remove': 'entfernen',
    'added': 'hinzugefügt',
    'add': 'hinzufügen',
    'Choose topics and courses for class': 'Auswahl von Themen und Kursen',
    'Group name': 'Name der Gruppe',
    'Year': 'Kalenderjahr',
    'Age': 'Alter der Teilnehmer',
    'In what setting will you teach the courses?': 'In welchem Rahmen findet der Kurs statt?',
    'Where will the class take place': 'In welchem Rahmen findet der Kurs statt',
    'School class': 'In der Schule',
    'Extracurricular': 'Außerschulisch',
    'Choose class type': 'Geplante Nutzung',
    'Add class information': 'Informationen zur Klasse',
    'Choose courses': 'Kursauswahl',
    'Available when all lessons are completed.': 'Verfügbar sobald alle Lektionen als erledigt markiert wurden.',
    'Click here to download your template.': 'Klicke hier um die Zertifikatvorlage herunterzuladen.',
    'Click to get help': 'Klicke um zur Hilfe zu gelangen',
    'Lesson content': 'Inhalte der Lektion',
    'Common mistakes': 'Häufige Fehler',
    'Lesson links & hints': 'Links & Hinweise zur Lektion',
    'previous lesson': 'vorherige Lektion',
    'next lesson': 'nächste Lektion',
    'completed {completedLessons} of {totalLessons} lessons': '{completedLessons} von {totalLessons} Lektionen erledigt',
    'Enter school type': 'Schultyp eingeben…',
    'School type': 'Schulart',
    'School name': 'Schulname',
    'Subjects': 'Fächer',
    'Profession / professional background': 'Beruf bzw. beruflicher Hinterund',
    'Enter country': 'Land eingeben...',
    'Canton': 'Kanton',
    'Federal state': 'Bundesland',
    'State': 'Bundesstaat',
    'District': 'Distrikt',
    'Region': 'Region',
    'Choose {entity}': '{entity} auswählen',
    'Country': 'Land',
    'Postal code': 'Postleitzahl',
    'City': 'Stadt',
    'Clear all': 'Alle löschen',
    'Clear': 'Löschen',
    'Add label': 'Label hinzufügen',
    'No results found': 'Keine Ergebnisse',
    'Please select': 'Bitte auswählen',
    'Type to search': 'Tippen um zu suchen',
    'Add {label}': '{label} auswählen',
    'Preparation': 'Vorbereitung',
    'Course offering': 'Kursangebot',
    'There are no preparation materials for this topic.': 'Für dieses Thema sind keine Materialien zur Kursvorbereitung vorhanden.',
    'Here are important informations and materials to prepare your lessons.': 'Hier findest du wichtige Informationen und Unterlagen zur Vorbereitung.',
    'Elementary school': 'Grundschule',
    'Main school': 'Hauptschule',
    'Middle school': 'Realschule',
    'Comprehensive school': 'Gesamtschule / Gemeinschaftschule / Stadtteilschule',
    'Academic high school': 'Gymnasium',
    'Vocational school': 'Berufliche Schule',
    'University': 'Hochschule / Universität',
    'Special school': 'Förderschule',
    'Login successful. You will now be redirected...': 'Login erfolgreich. Sie werden jetzt weitergeleitet.',
    'Too many requests. Please try again in a few seconds.': 'Zu viele Anfragen. Bitte versuchen sie es in ein paar Sekunden erneut.',
    'There was an Error processing your request. Please contact philipp@appcamps.de if the problem persists.': 'Leider ist ein Fehler aufgetreten. Wende dich bitte an philipp@appcamps.de falls das Problem weiterhin besteht.',
    'Required': 'Erforderlich',
    'Invalid': 'Ungültig',
    'Not an email': 'Keine E-Mail',
    'Not equal': 'Nicht identisch',
    'You need to confirm our privacy policy': 'Du musst die Datenschutzbedingungen bestätigen',
    'At least 8 chars': 'Mindestens 8 Zeichen',
    'has already been taken': 'Bereits vorhanden',
  },
}
