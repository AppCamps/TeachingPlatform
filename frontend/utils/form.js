export function isNumber(value, previousValue) {
  if (!value) {
    return '';
  }
  const parsedValue = parseInt(value, 10);
  if (isNaN(parsedValue)) return previousValue;
  return parsedValue;
}

export function isGreaterThan(min) {
  return (value, previousValue) => {
    if (!value) return '';
    const parsedValue = isNumber(value, previousValue);
    return parsedValue > min ? parsedValue.toString() : previousValue;
  };
}

export function isLowerThan(max) {
  return (value, previousValue) => {
    if (!value) return '';
    const parsedValue = isNumber(value, previousValue);
    return parsedValue < max ? parsedValue.toString() : previousValue;
  };
}

export function isInRange(min, max) {
  return (value, previousValue) => {
    if (!value) return '';

    const parsedValue = isNumber(value, previousValue);
    return parsedValue > min - 1 && parsedValue < max + 1 ? parsedValue.toString() : previousValue;
  };
}
