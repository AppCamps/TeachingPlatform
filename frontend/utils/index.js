import originalSlugify from "slugify";

export function slugify(originalString) {
  if (!originalString) {
    return null;
  }
  const stringForSlug = originalString.replace(/[^A-Za-z0-9_\s]/g, "");
  return originalSlugify(stringForSlug.toLowerCase());
}

export function lowerCaseFirstLetter(originalString) {
  if (!originalString) {
    return null;
  }
  return originalString.charAt(0).toLowerCase() + originalString.slice(1);
}

export function cleanRegExpString(originalString) {
  const stringCopy = originalString;
  if (stringCopy.startsWith("/") && stringCopy.endsWith("/")) {
    return stringCopy.slice(1, stringCopy.length - 1);
  }
  return stringCopy;
}

const internalCardTester = new RegExp(
  /^https?:\/\/.*teach.*\.appcamps\.de\/karten\/.*/
);

/*
 * Checks if link is an internal link to embedded cards.
 * @param link:string
 * @returns boolean
 */
export function isInternalCardLink(link) {
  return internalCardTester.test(link);
}

/*
 * Transforms http(s)://.*teach.*.appcamps.de/karten/code123
 * into
 * {
 *    url: '.*teach.*.appcamps.de/karten',
 *    code: 'code123'
 * }
 * @param link:string
 * @returns { url: string, code: string }
 */
export function parseInternalCardLink(link) {
  const code = link.split("/karten/")[1];
  const url = link.replace(`/${code}`, "").replace(/^https?:\/\//, "");
  return {
    code,
    url,
  };
}
