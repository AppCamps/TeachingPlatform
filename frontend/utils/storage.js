
export class LocalStorage {

  /**
   * @param w - window object
   * */
  constructor(w) {
    if (!LocalStorage.isSupported(w)) {
      throw new Error('LocalStorage is not supported');
    }
    this.storage = w.localStorage;
  }

  /**
   * Check is localStorage/sessionStorage is supported.
   *
   * @param w - window object
   * */
  static isSupported(w) {
    return !!w.Storage;
  }

  set(key, value) {
    this.storage.setItem(key, value);
  }

  get(key) {
    return this.storage.getItem(key);
  }

  remove(key) {
    this.storage.removeItem(key);
  }

  clear() {
    this.storage.clear();
  }
}

export class CookieStorage {

  /**
   * @param w - window object
   * */
  constructor(w) {
    if (!CookieStorage.isSupported(w)) {
      throw new Error('Cookies are disabled');
    }
    this.doc = w.document;
  }

  /**
   * Check if cookies are enabled in the browser.
   * @param w - window object
   * */
  static isSupported(w) {
    return w.navigator.cookieEnabled;
  }

  set(key, value) {
    this.doc.cookie = `${key}=${value};path=/`;
  }

  get(key) {
    const cookieKey = `${key}=`;
    const keyValues = decodeURIComponent(this.doc.cookie).split(';');
    const match = keyValues.filter(keyValue => keyValue.startsWith(cookieKey));
    if (match.length === 0) {
      return undefined;
    }

    // Remove "key="
    return match[0].substring(cookieKey.length, match[0].length);
  }

  remove(key) {
    this.doc.cookie = `${key}=;expires=Thu, 01 Jan 1970 00:00:01 GMT;path=/`;
  }

  clear() {
    const keyValues = decodeURIComponent(this.doc.cookie).split(';');
    keyValues
      // Get keys from key value strings
      .map(keyValue => keyValue.split('=')[0])
      // Remove all keys
      .forEach(key => this.remove(key));
  }
}

export class MemoryStorage {

  /**
   * @param w - window object
   * */
  constructor() {
    this.storage = {};
  }

  static isSupported() {
    return true;
  }

  set(key, value) {
    this.storage[key] = value;
  }

  get(key) {
    return this.storage[key];
  }

  remove(key) {
    delete this.storage[key];
  }

  clear() {
    this.storage = {};
  }
}

/**
 * Obtain a functional storage.
 * */
export function getStorage(windowRef) {
  // Use window by default if not provided.
  // Giving it in param allows testing.
  const w = windowRef || window;
  try {
    return new LocalStorage(w);
  } catch (e) {
    // Fallback to next
  }
  try {
    return new CookieStorage(w);
  } catch (e) {
    // Fallback to next
  }
  // Always works
  return new MemoryStorage(w);
}

// Export functional storage
export default getStorage();
