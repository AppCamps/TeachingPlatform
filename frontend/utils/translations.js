import { translations } from "../translations";

let getState = null;
export function intializeTranslationUtils(store) {
  getState = store.getState;
}

export function t(templateKey, interpolationObject) {
  const { lang } = getState ? getState().i18nState : "de";

  let templateString = templateKey;
  if (translations[lang] && translations[lang][templateKey]) {
    templateString = translations[lang][templateKey];
  }

  const matches = templateString.match(/{([A-z0-9]*)}/g);
  if (!matches || matches.length === 0) {
    return templateString;
  }

  return matches.reduce((memo, match) => {
    const key = match.replace(/[{}]/g, "");
    const replaceString = interpolationObject[key];

    if (!replaceString) {
      throw new Error(
        `Could not translate '${templateString}' missing replacement for ${key}.`
      );
    }

    return memo.replace(new RegExp(match), replaceString);
  }, templateString);
}

function formErrors() {
  return {
    required: t("Required"),
    invalid: t("Invalid"),
    notAnEmail: t("Not an email"),
    notEqual: t("Not equal"),
    privacyPolicyNotAccepted: t("You need to confirm our privacy policy"),
    minimumCharCount: t("At least 8 chars"),
    alreadyExists: t("has already been taken"),
  };
}

export function translatedFormError(key, interpolationObject) {
  const translationTemplateString = formErrors()[key];
  if (!translationTemplateString) {
    throw new Error(`Translation for ${key} not found.`);
  }

  return t(translationTemplateString, interpolationObject);
}
