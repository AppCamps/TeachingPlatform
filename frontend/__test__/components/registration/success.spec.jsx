import React from 'react';
import PropTypes from 'prop-types';

import { mount } from 'enzyme';

import { expect } from '../../chai_helper';

import { t } from '../../../utils/translations';

import RegistrationSuccess from '../../../components/registration/success';

describe('<Registration />', () => {
  let redirectToRegistrationPageMock;
  beforeEach(() => {
    redirectToRegistrationPageMock = jest.fn();
  });

  const setupWrapper = props => (
    mount(
      <RegistrationSuccess
        {...{
          redirectToRegistrationPage: redirectToRegistrationPageMock,
          ...props,
        }}
      />,
      {
        context: {
          t,
        },
        childContextTypes: { t: PropTypes.func },
      },
    )
  );

  it('renders correctly with user firstName', () => {
    const wrapper = setupWrapper({ firstName: 'Thomas' });

    expect(wrapper).to.contain.text('Thomas');
    global.expect(wrapper).toMatchSnapshot();
  });

  it('redirects to Registration and renders nothing if no user is given', () => {
    const wrapper = setupWrapper({ firstName: null });

    expect(wrapper.children()).to.have.length(0);
    expect(redirectToRegistrationPageMock.mock.calls).to.have.length(1);
  });
});
