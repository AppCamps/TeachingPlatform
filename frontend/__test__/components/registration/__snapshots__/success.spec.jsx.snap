// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Registration /> renders correctly with user firstName 1`] = `
<RegistrationSuccess
  firstName="Thomas"
  redirectToRegistrationPage={[Function]}
>
  <div
    className="container"
  >
    <div
      className="registration"
    >
      <h1
        className="heading"
      >
        Registration success
      </h1>
      <div
        className="signInLink"
      >
        <Link
          onlyActiveOnIndex={false}
          style={Object {}}
          to="/login"
        >
          <a
            onClick={[Function]}
            style={Object {}}
          >
            Sign in
          </a>
        </Link>
      </div>
      <div
        className="registrationSuccess"
      >
        <div
          className="registrationIcon"
        >
          <Check
            color="white"
          >
            <svg
              viewBox="-2232.818 -1235.338 10.293 9.852"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1383,1407.041l3.234,4.506,6.247-8.7"
                fill="none"
                stroke="white"
                strokeWidth="1px"
                transform="translate(-3615.412 -2637.89)"
              />
            </svg>
          </Check>
        </div>
        <div
          className="confirmationInfo"
        >
          <h2>
            Welcome to App Camps, Thomas!
          </h2>
          <p>
            You will soon receive an confirmation email. To complete your registration, please click on the confirmation link in the email. Afterwards you will be able to sign in.
          </p>
          <p>
            If the email does not arrive within the next minutes please [object Object] it or [object Object].
          </p>
        </div>
      </div>
    </div>
  </div>
</RegistrationSuccess>
`;
